plugins/glustershd/transactions.go

```diff
@@ -0,0 +1,119 @@
+package glustershd
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/gluster/glusterd2/glusterd2/brick"
+	"github.com/gluster/glusterd2/glusterd2/daemon"
+	"github.com/gluster/glusterd2/glusterd2/servers/sunrpc/dict"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const syncMgmtOperation = "20"
+
+func getHxlChildrenCount(volinfo *volume.Volinfo) (int, string) {
+	if volinfo.Type == volume.Replicate || volinfo.Type == volume.DistReplicate {
+		return volinfo.Subvols[0].ReplicaCount, "replicate"
+	}
+	return volinfo.Subvols[0].DisperseCount, "disperse"
+}
+
+func addHxlatorToDict(reqDict map[string]string, volinfo *volume.Volinfo, index int, count int, xlType string) map[string]string {
+	key := fmt.Sprintf("xl-%d", count)
+	xName := fmt.Sprintf("%s-%s-%d", volinfo.Name, xlType, index)
+	reqDict[key] = xName
+	reqDict[xName] = strconv.Itoa(index)
+	return reqDict
+}
+
+func selectHxlatorsWithBricks(volinfo *volume.Volinfo, healType string) map[string]string {
+	index := 1
+	hxlatorCount := 0
+	reqDict := make(map[string]string)
+	reqDict["heal-op"] = healType
+	reqDict["xl-op"] = reqDict["heal-op"]
+	reqDict["volname"] = volinfo.Name
+	reqDict["sync-mgmt-operation"] = syncMgmtOperation
+	reqDict["vol-id"] = volinfo.ID.String()
+	hxlChildren, xlType := getHxlChildrenCount(volinfo)
+	localBricks := volinfo.GetBricks()
+	for brick := range localBricks {
+		hostKey := fmt.Sprintf("%d-hostname", index-1)
+		brickPathKey := fmt.Sprintf("%d-path", index-1)
+		reqDict[hostKey] = localBricks[brick].Hostname
+		reqDict[brickPathKey] = localBricks[brick].Path
+
+		if index%hxlChildren == 0 {
+			reqDict = addHxlatorToDict(reqDict, volinfo, (index-1)/hxlChildren, hxlatorCount, xlType)
+			hxlatorCount++
+		}
+		index++
+	}
+	reqDict["count"] = strconv.Itoa(hxlatorCount)
+	return reqDict
+}
+
+func txnSelfHeal(c transaction.TxnCtx) error {
+
+	var volname string
+	var volinfo volume.Volinfo
+	var healType string
+	if err := c.Get("volname", &volname); err != nil {
+		c.Logger().WithError(err).WithField(
+			"key", "volname").Error("failed to get value for key from context")
+		return err
+	}
+
+	if err := c.Get("volinfo", &volinfo); err != nil {
+		c.Logger().WithError(err).WithField(
+			"key", "volinfo").Error("failed to get value for key from context")
+		return err
+	}
+
+	if err := c.Get("healType", &healType); err != nil {
+		c.Logger().WithError(err).WithField(
+			"key", "healType").Error("failed to get value for key from context")
+		return err
+	}
+
+	glustershDaemon, err := newGlustershd()
+	if err != nil {
+		return err
+	}
+
+	c.Logger().WithFields(log.Fields{"volume": volname}).Info("Starting Heal")
+
+	client, err := daemon.GetRPCClient(glustershDaemon)
```

Comment:
I'm not sure how this will work. You are using the UUID of glusterd to form a path in `SocketFile()`

Please check by logging or attaching gdb to self hael daemon.